### ISSUE TITLE
Elm packages won't install to generated project

### ISSUE THREAD

**Issue by DigitalMachinist** [Gizra/generator-elmlang#31]
I'm basically a newb to Elm, but I've been in the market to move up to some better tooling for my Elm development, and this generator seemed like a great way to get going with that.

So far when using this generator I've encountered nothing but problems with running `gulp`, specifically because the project's Elm packages aren't being installed. For some reason, the generator doesn't install them and I can't manually install them with `elm package install`, either.

I tried starting a new Elm project on my machine with `elm package install ` and everything went fine, so I doubt that my machine's configuration is an issue here.

I've captured some terminal output to showcase the problem. As you can see, gulp fails as soon as it needs an Elm package. It recommends I remove elm-stuff and reinstall packages, so I do. `elm package install` reports that packages have been installed successfully (but takes a suspiciously too short period of time to complete). I run gulp again only to encounter the same problem.

Take a look:

```bash
$ gulp
(node:9336) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
[12:50:57] Using gulpfile ~\Documents\Tests\elm\elm-cube-thing\gulpfile.js
[12:50:57] Starting 'build'...
[12:50:57] Starting 'clean:dev'...
[12:50:57] Starting 'watch'...
[12:50:57] Finished 'watch' after 14 ms
[12:50:57] Finished 'clean:dev' after 30 ms
[12:50:57] Starting 'styles'...
[12:50:57] Starting 'copy:bower'...
[12:50:57] Starting 'elm-init'...
[12:50:57] Finished 'copy:bower' after 34 ms
[12:50:57] Starting 'copy:dev'...
[12:50:57] 'elm-init' errored after 129 ms
[12:50:57] Error in plugin 'gulp-elm'
Message:
    Could not find package evancz/elm-http.

Maybe your elm-stuff/ directory has been corrupted? You can usually fix stuff
like this by deleting elm-stuff/ and rebuilding your project.

[12:50:57] styles all files 67 B
[12:50:57] Finished 'styles' after 406 ms
[12:50:57] index.html & CNAME all files 292 B
[12:50:57] Finished 'copy:dev' after 109 ms

Jeff@Rose-Home-PC MINGW64 ~/Documents/Tests/elm/elm-cube-thing
$ rm -rf elm-stuff

Jeff@Rose-Home-PC MINGW64 ~/Documents/Tests/elm/elm-cube-thing
$ elm package install
Packages configured successfully!

Jeff@Rose-Home-PC MINGW64 ~/Documents/Tests/elm/elm-cube-thing
$ gulp
(node:14816) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
[12:55:27] Using gulpfile ~\Documents\Tests\elm\elm-cube-thing\gulpfile.js
[12:55:27] Starting 'build'...
[12:55:27] Starting 'clean:dev'...
[12:55:27] Starting 'watch'...
[12:55:27] Finished 'watch' after 15 ms
[12:55:27] Finished 'clean:dev' after 31 ms
[12:55:27] Starting 'styles'...
[12:55:27] Starting 'copy:bower'...
[12:55:27] Starting 'elm-init'...
[12:55:27] Finished 'copy:bower' after 35 ms
[12:55:27] Starting 'copy:dev'...
[12:55:27] 'elm-init' errored after 131 ms
[12:55:27] Error in plugin 'gulp-elm'
Message:
    Could not find package evancz/elm-http.

Maybe your elm-stuff/ directory has been corrupted? You can usually fix stuff
like this by deleting elm-stuff/ and rebuilding your project.
```

If it has any impact, I'm running a Windows 10 x64 install, and I'm using git-bash through ConEmu for my terminal.

I'm excited to make use of this generator. It really seems like the most fully-featured of the few that are available. If you happen to have a workaround for this issue I'll let you know how it pans out for me.

Thanks!

**Comment by amitaibu** [comment]
> because the project's Elm packages aren't being installed.

My first suspect is your Elm version. Can you confirm it's `v0.17`?

**Comment by DigitalMachinist** [comment]
Can do:

```bash
$ npm install -g elm@0.17.0
C:\Users\Jeff\AppData\Roaming\npm\elm-make -> C:\Users\Jeff\AppData\Roaming\npm\node_modules\elm\binwrappers\elm-make
C:\Users\Jeff\AppData\Roaming\npm\elm -> C:\Users\Jeff\AppData\Roaming\npm\node_modules\elm\binwrappers\elm C:\Users\Jeff\AppData\Roaming\npm\elm-reactor -> C:\Users\Jeff\AppData\Roaming\npm\node_modules\elm\binwrappers\elm-reactor
C:\Users\Jeff\AppData\Roaming\npm\elm-repl -> C:\Users\Jeff\AppData\Roaming\npm\node_modules\elm\binwrappers\elm-repl
C:\Users\Jeff\AppData\Roaming\npm\elm-package -> C:\Users\Jeff\AppData\Roaming\npm\node_modules\elm\binwrappers\elm-package

> elm@0.17.0 install C:\Users\Jeff\AppData\Roaming\npm\node_modules\elm
> node install.js

Downloading Elm binaries from https://dl.bintray.com/elmlang/elm-platform/0.17.0/win32-x64.tar.gz
C:\Users\Jeff\AppData\Roaming\npm
`-- elm@0.17.0


Jeff@Rose-Home-PC MINGW64 ~/Documents/Tests/elm/elm-cube-thing
$ gulp
(node:30792) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
[13:19:04] Using gulpfile ~\Documents\Tests\elm\elm-cube-thing\gulpfile.js
[13:19:04] Starting 'build'...
[13:19:04] Starting 'clean:dev'...
[13:19:04] Starting 'watch'...
[13:19:04] Finished 'watch' after 16 ms
[13:19:04] Finished 'clean:dev' after 34 ms
[13:19:04] Starting 'styles'...
[13:19:04] Starting 'copy:bower'...
[13:19:04] Starting 'elm-init'...
[13:19:04] Finished 'copy:bower' after 35 ms
[13:19:04] Starting 'copy:dev'...
[13:19:04] 'elm-init' errored after 127 ms
[13:19:04] Error in plugin 'gulp-elm'
Message:
    Could not find package evancz/elm-http.

Maybe your elm-stuff/ directory has been corrupted? You can usually fix stuff
like this by deleting elm-stuff/ and rebuilding your project.
```

**Comment by amitaibu** [comment]
Lets try this hint "Maybe your elm-stuff/ directory has been corrupted? You can usually fix stuff
like this by deleting elm-stuff/ and rebuilding your project."

I also see - `Could not find package evancz/elm-http.`, maybe we're using an old package here.

**Comment by DigitalMachinist** [comment]
Hmm.

The elm-package.json installed by the generator is practically identical to what I would consider a reference package.json for an elm v0.17 app at [https://github.com/elm-lang/package.elm-lang.org/blob/master/elm-package.json](https://github.com/elm-lang/package.elm-lang.org/blob/master/elm-package.json). The evancz/elm-http version information is the same, anyway.

**Comment by DigitalMachinist** [comment]
I took the statement `Could not find package evancz/elm-http.` to refer to the fack that my `elm-stuff/packages` folder is empty and `elm-stuff/exact-dependencies.json' is nothing by an empty JSON object even after running `elm package install`.

**Comment by amitaibu** [comment]
Can you confirm if this repo is working for you - https://github.com/Gizra/elm-spa-example

**Comment by amitaibu** [comment]
^^ it was also scaffolded from this generator

**Comment by DigitalMachinist** [comment]
> ^^ it was also scaffolded from this generator

Haha well that explains the similarity!

> Can you confirm if this repo is working for you - https://github.com/Gizra/elm-spa-example

No dice, but it's a different error with a different package.

```bash
C:\Users\Jeff\Documents\Tests\elm\elm-spa-example>gulp
(node:28524) fs: re-evaluating native module sources is not supported. If you are using the graceful-fs module, please update it to a more recent version.
[13:39:33] Using gulpfile ~\Documents\Tests\elm\elm-spa-example\gulpfile.js
[13:39:33] Starting 'build'...
[13:39:33] Starting 'clean:dev'...
[13:39:33] Starting 'watch'...
[13:39:34] Finished 'watch' after 23 ms
[13:39:34] Finished 'clean:dev' after 31 ms
[13:39:34] Starting 'styles'...
[13:39:34] Starting 'copy:bower'...
[13:39:34] Starting 'elm-init'...
[13:39:34] Finished 'copy:bower' after 35 ms
[13:39:34] Starting 'copy:dev'...
[13:39:34] styles all files 401 B
[13:39:34] Finished 'styles' after 601 ms
[13:39:34] index.html & CNAME all files 540 B
[13:39:34] Finished 'copy:dev' after 120 ms
[13:39:34] 'elm-init' errored after 159 ms
[13:39:34] Error in plugin 'gulp-elm'
Message:
    There are no versions of package 'elm-community/json-extra' on your computer.
```

**Comment by amitaibu** [comment]
I've just tried a fresh install of the SPA example. I did see some npm warning about deprecated packages, but no errors, and `gulp` is working.

```
elm-package install -y
npm i
```

elm-package: 0.17.1
node: v4.2.2

**Comment by DigitalMachinist** [comment]
It might be notable that I'm running the latest Node v6.2.2. I don't have nvm installed on this machine, so testing other versions might be inconvenient.

**Comment by amitaibu** [comment]
Yeah, I'm pretty sure it's somehow related to your environment. Do you have another computer you could check on?

We've already been using this generator on several different machines

**Comment by DigitalMachinist** [comment]
I'll dig something up to try. Not certain if my laptop has elm installed, but I can take a stab at it if so.

I'll let you know what I discover.

**Comment by DigitalMachinist** [comment]
So it appears that my laptop gives me the very same error I originally reported when trying to run gulp on a generated project, and the same error I got on this machine upon cloning elm-spa-example. It's also on Windows 10 x64, but is running Node v5.3.0. I made certain its Elm version was up to date.

I mean, I configured both of these machines so it's possible I have the same misconfiguration in 2 places, but this does look a little bit fishy to me,

**Comment by amitaibu** [comment]
> Windows 10 

Admittedly I've always tried it on Mac/ Ubuntu. Did it install the Elm packages? If not, can you paste the content of it here please

**Comment by DigitalMachinist** [comment]
I just started from scratch. My process here was as follows:

```bash
mkdir elm-cube-thing
cd elm-cube-thing
yo elmlang
gulp
```

At which point it fails as usual. At this point, my elm-stuff folder looks like this:

```
elm-cube-thing
  - elm-stuff
    - exact-dependencies.json
    - packages
      - (empty)
```

Where exact-dependencies.json contains:

```json
{}
```

This is the same state it is in after a remove it and run `elm package install`.

**Comment by amitaibu** [comment]
I think your case is not necessarily related to the generator. Are you able to install a new elm project form scratch (without this generator)?

**Comment by DigitalMachinist** [comment]
Well, I'm feeling pretty dumb right now.

I forgot that Elm has a standalone installer, and because the elm command doesn't have a version option, I didn't think I had a way of directly confirming the vevrsion. I thought my `npm install -g elm@0.17.0` was evidence enough to think that v0.17.0 was installed, but I was wrong. Turns out the standalone installer is the olny way.

I discovered my error when I mistyped an elm command and it said v0.16.0 right there in front of me.

I've just run the standalone installer for v0.17.1 on my machine and it looks like things are fine.

Sorry to waste so much of your time on such a trivial issue (that you obviously spotted right at the start)... Argh.

Thanks for all of your help ironing this out, though. I really appreciate it.

**Comment by amitaibu** [comment]
No worries, It at least confirms that I was spot on in my [first comment](https://github.com/Gizra/generator-elmlang/issues/31#issuecomment-231145015) 😉

**Comment by amitaibu** [comment]
Actually if you'd like to "pay for your sins" (joking of course), a PR to confirm a computer has the right version could be really cool. I'd be happy to guide you through it.

**Comment by DigitalMachinist** [comment]
I can take a stab at it. I've work on another (fairly trivial) yo generator before so I know some of the basics, at least.

**Comment by DigitalMachinist** [comment]
Basically all I really need to know is where you want to fit the version check into the code.

I figure I'm probably just going to be executing the elm command from JS and then testing the stdout output with a regex for the version number, because I don't know of any other way to get the version data from an elm install, and it seems like the most reliable way.

**Comment by amitaibu** [comment]
Cool. 

I think we can have a `validateEnv` phase before the [install](https://github.com/Gizra/generator-elmlang/blob/master/generators/app/index.js#L126) where we could [throw an error](https://github.com/Gizra/generator-elmlang/blob/master/generators/app/index.js#L36) if Elm isn't v0.17.x

**Comment by DigitalMachinist** [comment]
All right. PR #32 has been submitted.

So I ended up going a little further than a regex test, because when I got down to it, that seemed like a real maintenance headache down the road. Having to tweak a regex to handle a list of supported semvers or a range of semvers would become a strain real fast.

So I did you one better.

I'm reading the valid version range out of `./templates/_elm-platform.json` and parsing it for the minimum version and the earliest unsupported version, then checking if the installed elm version reported by invoking `elm` on the console falls inside that range. I've included the `semver` package from npm to do this is a robust way without having to code up a semver check myself.

Hopefully this saves people some trouble in the future!
